Trell.Test/Engine/EngineTest.cs

```diff
@@ -1,49 +1,160 @@
-        while (true) {
-            var val = rng.Next();
-            var name = $"trell-test-{val}";
-            var dir = Path.GetFullPath(Path.Combine(Path.GetTempPath(), name));
-            if (!Directory.Exists(dir)) {
-                Directory.CreateDirectory(dir);
-                tmpDir = dir;
-                break;
+    }
+
+    void WriteWorkerJsFile(string filename, string? scheduled = null, string? fetch = null, string? upload = null) {
+        File.WriteAllText(
+            Path.GetFullPath(filename, this.EngineDir),
+            $$"""
+            async function scheduled(event, env, ctx) {
+                {{scheduled}}
+            }
+            
+            function fetch(request, env, ctx) {
+                {{fetch}}
+            }
+            
+            function upload(payload, env, ctx) {
+                {{upload}}
             }
+            
+            export default {
+                scheduled,
+                fetch,
+                upload,
+            }
+
+            """
+        );
+    }
+
+    public void Dispose() {
+        if (this.disposed) {
+            return;
+        }
+
+        this.disposed = true;
+        GC.SuppressFinalize(this);
+
+        if (Directory.Exists(this.EngineDir)) {
+            Directory.Delete(this.EngineDir, true);
+        }
+    }
+
+    ~EngineFixture() => Dispose();
+}
+
+public class EngineTest(EngineFixture engineFixture) : IClassFixture<EngineFixture> {
+    EngineFixture fixture = engineFixture;
+
+    [Fact]
+    public async Task TestEngineWrapperOnlyRunsPredeterminedCommands() {
+        var eng = MakeNewEngineWrapper();
+        var ctx = MakeNewExecutionContext();
+
+        var work = new Work(new(), "{}", this.fixture.EngineDir, "NotAValidFunctionName");
+        await Assert.ThrowsAsync<TrellUserException>(async () => await eng.RunWorkAsync(ctx, work));
+
+        string[] validFunctions = ["scheduled", "upload", "fetch"];
+        foreach (var function in validFunctions) {
+            work = work with { Name = function };
+            var x = await eng.RunWorkAsync(ctx, work);
+            Assert.NotNull(x);
         }
+    }
+
+    [Fact]
+    public async Task TestEngineWrapperCorrectlyCreatesJsBuffers() {
+        var eng = MakeNewEngineWrapper();
+        var ctx = MakeNewExecutionContext();
+
+        bool parsed = TrellPath.TryParseRelative("js_buffer_checking_worker.js", out var workerPath);
+        Assert.True(parsed);
+
+        byte[] original = [ 0x03, 0x51, 0x44, 0x3A, 0xC5 ];
+        var buffer = eng.CreateJsBuffer(original);
+        Assert.NotNull(buffer);
+
+        var work = new Work(new(), "{}", this.fixture.EngineDir, "upload") {
+            WorkerJs = workerPath!,
+            Arg = new Work.ArgType.Raw(buffer),
+        };
+        var actual = await eng.RunWorkAsync(ctx, work);
+        Assert.Equal("true", actual);
+    }
+
+    [Fact]
+    public async Task TestEngineWrapperCanTimeOut() {
```

Comment:
This tests that the timeout works in the JS code - which is great. I think we also want a test that ensure the timeout works in the C# code. I can explain tomorrow.

Reply:
Well, actually it calls `console.log` so it could be either. Might be worth splitting that out.